### ISSUE TITLE
Move the serializing component (TypeAdapter and hive_generator) to its own package?

### ISSUE THREAD

**Issue by marcelgarus** [hivedb/hive#152]
There are several cases where you need to serialize data. For example, there's storage (what hive does), but also communication over a network or between `Isolate`s where the serialization of data makes sense. My point is, serialization is not a problem specific to Hive.

I like Hive's protobuf-inspired `TypeAdapter` concept with hive_generator. It's lightweight, forward-compatible and produces a small output (in contrast to, for example, the json_serialize package).

I'm wondering whether it's possible to release a package separate from Hive that only exposes this serializer? Then, Hive and other packages could depend on it, making improvements in the serializer available to everyone.

**Comment by leisim** [comment]
> Hive and other packages could depend on it, making improvements in the serializer available to everyone.

I like the idea of decoupling the `BinaryReader` and `BinaryWriter`.

> I'm wondering whether it's possible to release a package separate from Hive that only exposes this serializer?

It is possible but currently, both `BinaryReader` and `BinaryWriter` have Hive specific logic (like [`writeHiveList()`](https://github.com/hivedb/hive/blob/0370d564eb8cbd64010244aef39ecc712cdb020b/lib/src/binary/binary_writer_impl.dart#L231)).

I'm not sure whether to include the `Frame` class and the corresponding logic like [`write()`](https://github.com/hivedb/hive/blob/0370d564eb8cbd64010244aef39ecc712cdb020b/lib/src/binary/binary_writer_impl.dart#L273).

**Comment by marcelgarus** [comment]
Of course, naming is always a difficult topic. Maybe just something like `BinarySerializer`? Or some completely custom name?

But I didn't know the serialization part and the _saving things_-part are that tightly coupled. I assumed objects would be converted to an intermediate format, like a `Uint8List`, and then that would get written into storage.

**Comment by marcelgarus** [comment]
Okay, I did some thinking and I possibly found a solution.

The `BinaryReader` and `BinaryWriter` are currently monolithic classes with lots of methods, some of which are directly tied to Hive. I'll describe my proposal for the `BinaryReader`, the `BinaryWriter` works analogous.

The `BinaryReader`'s `readHiveList` method uses only the `readUint32`, `readByte`, `readAsciiString` and `readKey` methods instead of acting on the class's data directly. That means, the `readHiveList` method is at another layer of abstraction. That holds true for many of the methods – `readList`, `readMap` etc. all use simpler, more fundamental methods.

If we were to have different implementations of the `BinaryReader` (the Hive implementation, but also other ones for networking etc.), then it would make no sense for them to have to implement the `readHiveList` and `readList`, `readMap` etc. methods – rather they could simply implement the more fundamental methods and the more abstract ones live on top in another layer of abstraction.

If we would split the `BinaryReader` into a base class with only fundamental methods, like `readByte`, and `readBytes` plus `seek` and `skip`, the other methods could be converted into `TypeAdapter`s. For example, the `readHiveList` method could easily be implemented as a `TypeAdapter<HiveList>` and then get automatically registered during startup. The same could even hold true for `TypeAdapter<int>`, `TypeAdapter<String>` etc.
Different encoding variants of the same type like ascii strings and several types of `int`s could be added as extension methods on the original writer or put directly into the abstract class.

No change would be necessary for the code generator, because it can still use the `BinaryWriter` and `BinaryReader` as usual. Instead of using `readHiveList` though, it would simply use `read` and then cast the result to a `HiveList` – just like with any other custom type.

If implemented, here's how the code would be separated:

The new serializing package would contain the abstract `BinaryReader` and `BinaryWriter` including already implemented methods for reading/writing basic types like `Uint32`, `Map`s, `List`s, `Set`s etc. Everything from the hive_generator package would also move to a serializing generator package.

The `hive` package contains the `BinaryReaderImpl` and `BinaryWriterImpl` to actually read and write bytes to disk. Also, it implements and registers type adapters for hive-specific types (aka the `HiveList`).

Other packages could implement their own `BinaryReaderImpl` and `BinaryWriterImpl` to read/write bytes to the network or whatever.

How should we proceed? What do you think about this approach? And knowing that this has not the highest priority for you right now, should I start implementing it? It's currently a road blocker for a package I'm developing during my free time, so I'd be welcome to help.

**Comment by leisim** [comment]
Happy new year @marcelgarus 

I had a similar thought. The Hive specific logic (`readKey()`, `readFrame()` and `readHiveList()`) could be implemented using extension methods for `BinaryReader` and `BinaryWriter`.

But what about `readList()`, `readMap()` and `read()`?

> the readHiveList method could easily be implemented as a TypeAdapter<HiveList> and then get automatically registered during startup. The same could even hold true for TypeAdapter<int>, TypeAdapter<String> etc.

Yes, but it would impact performance. The `TypeAdapter` lookup is `O(n)` with n being the number of registered adapters.

Would you also put the `TypeAdapter` interface, the `TypeRegistry` and the `hive_generator` into the new package? The `TypeAdapter` is kind of Hive specific with the recent change (#163).
And the `hive_generator` depends heavily on Hive. Honestly, I have no idea how to create a generic build runner that can be extended by a package.

> How should we proceed? should I start implementing it?

Let's work out the details first and then we can figure out the implementation.

**Comment by marcelgarus** [comment]
> Yes, but it would impact performance. The TypeAdapter lookup is O(n) with n being the number of registered adapters.

Architecture-wise, it seems more elegant to register `TypeAdapter`s for `List` and `Map`, but for performance reasons, of course, it is reasonable to add `readList()` and `readMap()` as extension methods or implemented methods in the abstract class.
`read()` itself could also be an extension method or an implemented method in the abstract class. It depends on the `TypeRegistry` and `TypeAdapter`s, but these really should be part of the serializer, see below.  

> Would you also put the TypeAdapter interface, the TypeRegistry and the hive_generator into the new package?

I think, `TypeAdapter`s and the `TypeRegistry` as well as the `hive_generator` should also be part of the serializing package for several reasons.
They belong to the serializer semantically, because they care about serializing custom types.
Also, from the user's perspective, it would be much easier to just annotate the classes once to support binary serialization and _it just works_ in all libraries that depend on it, rather than having to annotate each class and field with multiple annotations for the different libraries.  
Not putting these classes into the serializing package, I feel like the benefit of a common serializing package gets greatly reduced.

> The TypeAdapter is kind of Hive specific with the recent change (#163).

I don't see how adding the `typeId` to to `TypeAdapter` makes the package any more Hive-specific. Could you elaborate on that?

> And the hive_generator depends heavily on Hive.

To be honest, I don't see that either. The `@HiveType` and `@HiveField` annotations would also be part of the serializing package (perhaps renamed?) and the `HiveList` can be encoded using a `TypeAdapter`. I _do_ admit, this impacts performance a bit.

---

By the way, I do realize this is a very long term and breaking change, so if you are hesitant or want to wait for other features to be adopted first, I get that.
In the meantime, I'll try to implement my own serializer and also play with some optimizations and features that I think will be interesting. For example, by determining the adapter during code generation rather than at runtime for statically typed fields, encoding of both subclasses and superclasses, as well as generic types, could be supported.

**Comment by leisim** [comment]
> Architecture-wise, it seems more elegant to register TypeAdapters for List and Map.

I agree. 

> I don't see how adding the `typeId` to to `TypeAdapter` makes the package any more Hive-specific.

Nevermind, I was confused for a second ^^

> the HiveList can be encoded using a TypeAdapter

Unfortunately, I don't think using `TypeAdapters` works. The `BinaryReader.read()` method returns `dynamic` due to Dart limitations. So the `hive_generator` needs to cast any list or map (including `HiveLists`). In theory, `HiveList` could override the `cast()` method from `List` but I could not make it work. The problem is that `HiveList` requires its elements to extend `HiveObject`.

```dart
@override
HiveList<T> cast<T extends HiveObject>() { // -> invalid override
  return HiveListImpl(box, objects: _delegate.cast());
}
```

Due to this limitation, `HiveList` needs a function called `castHiveList()` and the `hive_generator` calls this method.

> @HiveType and @HiveField annotations would also be part of the serializing package (perhaps renamed?)

Maybe `@BinaryType` and `@BinaryField`. Or we can find a way to make `hive_generator` (or whatever it is called then) customizable.

> By the way, I do realize this is a very long term and breaking change, so if you are hesitant or want to wait for other features to be adopted first, I get that.

I think "Hive 2.0" is not too far away and we need some time for a future proof implementation anyway. I have no problem making breaking changes if it improves the package.

TLDR: I agree with everything you said but we need to figure out a way to let packages extend the `hive_generator`.

**Comment by marcelgarus** [comment]
Hmm okay, I do see the problem with the `HiveList`. I'm not sure how to make the current `hive_generator` extensible either – I'm not sure it's even possible at all because all the code that runs during code generation is pre-defined.

Determining the adapter during code generation would solve the problem, so here are my thoughts on the topic so far:

**Determining the adapter during code generation**

If we would serialize objects this way, we retain type safety in statically typed contexts. We could rely on Adapter classes having a special name of the form `AdapterForType<GenericType>`. Because generated adapters live in the same library as the original class (the `.g.dart` file), classes that are in the namespace automatically have an adapter that's also in the namespace.
The syntax could look somewhat like this:

```dart
class AdapterForList<T> implements TypeAdapter<List<T>> {
  AdapterForList({@required this.adapterForT});

  final TypeAdapter<T> adapterForT;

  @override
  int get typeId => 4;

  void write(BinaryWriter writer, List<T> list) {
    writer.write(list.length, adapter: AdapterForCollectionLength());
    for (final item in list) {
      writer.write(item, adapter: adapterForT);
    }
  }

  List<T> read(BinaryReader reader) {
    final length = reader.read(adapter: AdapterForCollectionLength());
    return <T>[
      for (var i = 0; i < length; i++) reader.read(adapter: adapterForT),
    ];
  }
}

// A very simple generic class.
class Id<T> {
  Id(this.id);

  final int id;
}

// This adapter could omit the adapterForT by recognizing there's no field with the type.
class AdapterForId<T> implements TypeAdapter<Id<T>> {
  @override
  int get typeId => 10;

  @override
  Id<T> read(BinaryReader reader) {
    return Id<T>(reader.read(adapter: AdapterForInt()));
  }

  @override
  void write(BinaryWriter writer, Id<T> obj) {
    writer.write(obj.id);
  }
}

class MyClass<T> {
  MyClass({@required this.id, @required this.someNumbers});

  final Id<MyClass> id;
  final List<int> someNumbers;
}

class AdapterForMyClass<T> implements TypeAdapter<MyClass<T>> {
  AdapterForMyClass({@required this.adapterForT});

  final TypeAdapter<T> adapterForT;

  @override
  int get typeId => 11;

  @override
  MyClass<T> read(BinaryReader reader) {
    final id = reader.read(adapter: AdapterForId<MyClass>());
    final someNumbers = reader.read(
      adapter: AdapterForList<int>(
        adapterForT: AdapterForInt(),
      ),
    );
    return MyClass<T>(
      id: id,
      someNumbers: someNumbers,
    );
  }

  @override
  void write(BinaryWriter writer, MyClass<T> obj) {
    writer
      ..write(obj.id, adapter: AdapterForId())
      ..write(
        obj.someNumbers,
        adapter: AdapterForList<int>(adapterForT: AdapterForInt()),
      );
  }
}
```

Note that there are several great features that improve upon the current design:

- There is no adapter lookup of O(n) anymore.
- The `typeId` doesn't have to be saved explicitly anymore, because we know that the reader knows by context which type to expect. That also means types have _no storage overhead by default_. Currently, common cases like `List<String>` and `List<int>` have special encodings for more efficiency, this wouldn't be necessary anymore.
- Encoding both a supertype and a subtype would work as well. Currently, when encoding the subtype, two type adapters match and it's impossible for the serializer to choose.
- Generics work.

What still needs to be worked out is that this only works in statically typed contexts. Anywhere where a dynamic value is expected, we need to encode the type as well. That also applies to collections, because they may contain subclasses or implicit generic dynamics.

In the example, a `List<Id>` could be constructed, containing `[Id<int>(0), Id<String>(3), Id<MyClass>(7)]`. With some tweaking, retaining types in such a situation is possible, I think.

But there are limits to type safety: If we have a `List<dynamic>` with `[3, Id<T>(8)]`, to my understanding there is currently no way in Dart to find out what `T` is.

Also, forward compatibility is still an issue, because in the code above, we don't save field ids anymore. There's probably a simple solution to that, like first saving all the field ids along with an offset for each and then running the appropriate adapters at the offsets.

**About annotations**

> Maybe `@BinaryType` and `@BinaryField`.

I like `@BinaryType` and `@BinaryField`. If we find a name other than "Binary" for the serializer, prefixing with that would also be fine with me.

> Or we can find a way to make `hive_generator` (or whatever it is called then) customizable.

I'd strongly advise against that, because then for multiple libraries that use the binary serialization, you'd need multiple annotations on every field again.

**About keeping track of `typeId`s and `fieldId`s automatically**

For `typeId`s and `fieldId`s , dynamically generating them is obviously not a good option, as you pointed out at #163. The key part is that these IDs are not generated but explicitly mapped in the code.

So for `fieldId`s, an extension would perhaps be better suited for the job. For example, in VS Code, there could be a command for making the current class serializable, aka automatically annotating it with `@BinaryType` and enumerating the fields with `@BinaryField` annotations as well as adding the `part '...g.dart'` directive.
There could also be a remove command when the cursor is on a field, automatically inserting a comment `// don't use the following field ids for legacy reasons: 3` at the top of the class or appending to that list.
Every time, the first command is run, it could annotate fields that don't have an annotation yet, while respecting the comment at the top of the class.

For `typeId`s, a global map as proposed in #163 would probably be an option. It could be placed in a special `binary.dart` file or something similar and work just like the `fieldId` but on a class level.

**Automatically registering types?**

While we're at it and have a global map of types to their IDs, it's no harm to automatically register all these types, right? It would certainly reduce friction if you could just trigger a VS Code command to make a class serializable, run a command and it would just work out of the box without having to register any types.

**Breaking changes**

> I have no problem making breaking changes if it improves the package.

I agree with you here on a code level, but if we were to change the binary representation itself, we would not only bother developers but also break apps on the user's side. So, we would certainly need a transition path for that.

**Comment by leisim** [comment]
I like your idea but I have still some questions:

#### How will objects without generated adapter be saved?
The `adapterForT` parameter is a good idea but Hive also supports storing objects without generated adapter. How would storing a `List<SomeObject>` for example work? The user would need to register `AdapterForList<SomeObject>` manually.

#### Information of subtypes is lost
I can't think of a solution for the following:
```dart
class Pet { }

class Dog extends Pet {
  String name;
}

class MyClass {
  Pet pet;

  MyClass(this.pet);
}
```
If the user wants to save `MyClass(Dog())`, all the fields of `Dog` will be lost because the `AdapterForMyClass` expects a `Pet`.

#### Forward compatibility
> There's probably a simple solution to that, like first saving all the field ids along with an offset for each and then running the appropriate adapters at the offsets.

If I understand you correctly, you want to keep track of the position of the fields in a file or something similar. I don't think it is that easy: If the user changes the class multiple times, I don't see how we could track the field.

> For example, in VS Code, there could be a command for making the current class serializable, aka automatically annotating it with @BinaryType and enumerating the fields with @BinaryField annotations as well as adding the part '...g.dart' directive.

> While we're at it and have a global map of types to their IDs, it's no harm to automatically register all these types, right?

Sounds great :+1:

**Comment by marcelgarus** [comment]
After several prototyping attempts, I just realized the gravity of the fact that if we have a `List<T>` given to a function that accepts a `dynamic`, there is no way in Dart to find out what T is. This is really mind-boggling to me because it prevents us from writing a type-safe serializer (which you can give any object, like `List<T>` and it returns a representation, that if interpreted will return a `List<T>` again).

Grudgingly accepting that, I'm now more comfortable with dropping type information.

Regarding generics, I think my initial aspirations with statically resolved adapters are not feasible – as you said if you have a `List<Object>` that saves several numbers and strings, we don't need or want the calling side to be forced to specify an adapter for `Object`. I agree that the approach with `adapterForT` doesn't work in that case, so let's abandon that idea.

However, I'd still use generic types in the `AdapterFor…` type itself:

```dart
class AdapterForList<T> implements TypeAdapter<List<T>> {
  @override
  int get typeId => 4;

  void write(BinaryWriter writer, List<T> list) {
    writer.write(list.length, adapter: AdapterForCollectionLength());
    list.forEach(writer.write);
  }

  List<T> read(BinaryReader reader) {
    final length = reader.read(adapter: AdapterForCollectionLength());
    return <T>[
      for (var i = 0; i < length; i++) reader.read(),
    ];
  }
}
```

When a typed `List` is saved inside an object, it is read using `reader.read(adapter: AdapterForList<int>())`, providing a `List<int>`. In dynamic contexts just `reader.read()` is called and returns a `List<dynamic>`.

A class adapter would look like the following:

```dart
@BinaryType(11)
class MyClass<T> {
  MyClass({@required this.id, @required this.someNumbers});

  @BinaryField(0)
  final Id<MyClass> id;

  @BinaryField(1, defaultValue: <int>[])
  final List<int> someNumbers;
}

class AdapterForMyClass<T> implements TypeAdapter<MyClass<T>> {
  AdapterForMyClass();

  @override
  int get typeId => 11;

  @override
  MyClass<T> read(BinaryReader reader) {
    var numOfFields = reader.read(adapter: AdapterForNumberOfFields());
    var fields = <int, dynamic>{
      for (var i = 0; i < numOfFields; i++) reader.read(adapter: AdapterForFieldId()): reader.read(),
    };
    return MyClass(
      id: fields[0],
      someNumbers: fields.containsKey(1) ? fields[1] : <int>[],
    );
  }

  @override
  void write(BinaryWriter writer, MyClass<T> obj) {
    writer
      ..write(2)
      ..write(0)
      ..write(obj.id)
      ..write(1)
      ..write(obj.someNumbers, adapter: AdapterForList<int>());
  }
}
```

**Regarding your questions:**

> How will objects without generated adapter be saved?

Now, there is no more `adapterForT`, so that works.

> Information of subtypes is lost

With the current implementation, that also happens. More specifically, it only happens if both the superclass and subclass have adapters:

```dart
@BinaryType
class Pet {}

@BinaryType
class Dog extends Pet {
  String name;
}
```

That doesn't work with the current version either:

```dart
_ResolvedAdapter findAdapterForValue(dynamic value) {
  for (var adapter in _typeAdapters.values) {
    if (adapter.matches(value)) return adapter;
  }
  return null;
}
```

If the `PetAdapter` comes before the `DogAdapter` in `_typeAdapters.values`, the `PetAdapter` gets chosen instead of the `DogAdapter`.

Is there any disadvantage of directly instantiating an adapter?
Then we could register the adapters in a hierarchy.
For example, each adapter could have a collection of `subAdapters`:

```dart
class AdapterForPet extends TypeAdapter<Pet> {
  Set<TypeAdapter<Pet>> subAdapters;

  void registerSubAdapter(TypeAdapter<Pet> adapter) => subAdapters.add(adapter);

  // ...
}

const adapterForPet = AdapterForPet();
```

I already mentioned automatically registering adapters, so we could automatically generate code like the following:

```dart
// dog_adapter.g.dart

// ... the DogAdapter

const adapterForDog = AdapterForDog();

// The adapterForPet is in scope, because the Dog extends Pet, so Pet is in scope.
adapterForPet.registerSubAdapter(DogAdapter());
```

Then, the `findAdapterForValue` method could be implemented like this:

```dart
_ResolvedAdapter findAdapterForValue(dynamic value) {
  final TypeAdapter matchingAdapter;
  while (true) {
    for (var adapter in matchingAdapter?.subAdapters ?? _typeAdapters.values) {
      if (adapter.matches(value)) {
        matchingAdapter = adapter;
        break;
      }
    }
    // No (sub)adapter matched.
  }
  return matchingAdapter;
}
```

Note that because of the hierarchical structure, performance is somewhere between O(log n) and O(n) (depends on the class hierarchy), as opposed to always O(n) currently.

**Comment by marcelgarus** [comment]
I wasn't sure if the hierarchical type adapter structure would work, so I implemented a [proof of concept](https://github.com/marcelgarus/portals/tree/master/portals/lib/src/binary).

Here's an example class along with the serializer that could be generated:

```
class MyClass<T> {
  MyClass({@required this.id, @required this.someNumbers});

  final String id;
  final List<T> someNumbers;
}

class AdapterForMyClass<T> extends TypeAdapter<MyClass<T>> {
  const AdapterForMyClass();

  @override
  void write(BinaryWriter writer, MyClass<T> obj) {
    writer
      ..writeNumberOfFields(2)
      ..writeFieldId(0)
      ..write(obj.id)
      ..writeFieldId(1)
      ..write(obj.someNumbers);
  }

  @override
  MyClass<T> read(BinaryReader reader) {
    final numberOfFields = reader.readNumberOfFields();
    final fields = <int, dynamic>{
      for (var i = 0; i < numberOfFields; i++)
        reader.readFieldId(): reader.read(),
    };

    return MyClass<T>(
      id: fields[0],
      someNumbers: fields[1],
    );
  }
}
```

Regarding registering adapters and ownership of type ids, I currently don't have the id saved in the adapter. There are arguments for both sides, but eventually, type ids have to be managed across multiple data types and with generic types working, the same adapter can be registered using multiple ids.

Also, all the ids are offset when being encoded, so that built-in ids are now negative and users can use ids starting at 0.

When trying to serialize the type from above, an exception gets thrown:

```
No adapter for the type MyClass<int> found. Consider adding an adapter for that type by calling AdapterForMyClass<int>().registerWithId(0).
```

Doing that, another exception occurs and a warning occurs in the console:

```
No adapter for the exact type List<int> found, so we're encoding it as a List<dynamic>. For better performance and truly type-safe serializing, consider adding an adapter for that type by calling AdapterForList<int>().registerWithId(1).
```

Because the static type of a `List<int>` is not equal to the runtime type of `List<int>` (that means, `<int>[1,2,3].runtimeType != List<int>`), sadly another warning appears in the console, but the serializing works fine:

```
No adapter for the exact type List<int> found, so we're encoding it as a List<int>. For better performance and truly type-safe serializing, consider adding an adapter for that type by calling AdapterForList<int>().registerWithId(3).
Serialized to [8, 1, 0, 2, 0, 0, 7, 249, 0, 0, 0, 7, 115, 111, 109, 101, 45, 105, 100, 0, 1, 8, 2, 0, 0, 0, 3, 7, 250, 0, 0, 0, 0, 0, 0, 240, 63, 7, 250, 0, 0, 0, 0, 0, 0, 0, 64, 7, 250, 0, 0, 0, 0, 0, 0, 8, 64]
MyClass(some-id, [1, 2, 3])
```

```
void main() {
  AdapterForMyClass<int>().registerWithId(0);
  AdapterForList<int>().registerWithId(1);

  final data = serialize(MyClass(id: 'some-id', someNumbers: [1, 2, 3]));
  print('Serialized to $data');
  print(deserialize(data));
}
```

**Comment by leisim** [comment]
>  I just realized the gravity of the fact that if we have a List<T> given to a function that accepts a dynamic, there is no way in Dart to find out what T is

It is super annoying and was a major bummer when I implemented Hive. Unfortunately, we have to live with that.

> If the `PetAdapter` comes before the `DogAdapter` in `_typeAdapters.values`, the `PetAdapter` gets chosen instead of the `DogAdapter`.

Yes, that's correct and I seemingly forgot to document it. I'm thinking about using a `Map<Type, TypeAdapter>` to allow O(1) access if an adapter has the exact same type as a value. If there is no such adapter, the `TypeRegistry` will use the same approach as before to find the correct adapter.

I think the `subAdapters` hierarchy is a great idea to improve performance.

> Also, all the ids are offset when being encoded, so that built-in ids are now negative and users can use ids starting at 0.

Hive does that as well.

I'm still not sure whether I like the idea to create adapters for primitives and Lists. The reason is that the `BinaryReader` and `BinaryWriter` methods can access the byte buffer directly which improves performance. I don't see why anyone would want to override the default behavior for storing primitives.
In some cases it might be useful (for example in the Hive Sketchpad demo) but then implementing a custom `TypeAdapter` for the whole object is a better idea.

I've seen your implementation for `portals`. It looks very promising. A small note on using `ByteData` and `Uint8List`. Creating a new instance of either of them is rather expensive (even `Uint8List.view()`). You should consider replacing for example
```dart
void writeByte(int value) => writeByteData(ByteData(1)..setUint8(0, value));
```
with something that writes to the underlying buffer directly. I spent countless hours running microbenchmarks xD (the docs on improving performance of Dart programs is virtually non-existen)

**Comment by marcelgarus** [comment]
> I'm thinking about using a Map<Type, TypeAdapter> to allow O(1) access if an adapter has the exact same type as a value. If there is no such adapter, the TypeRegistry will use the same approach as before to find the correct adapter.

That's the exact approach I'm taking too. 😊 Constant lookup works fine with user-defined types, but types like `List` are very thoroughly engrained in the VM, so it doesn't work for them. Also, if the type is just an interface and the implementation is private, that doesn't work either (for example, `<int>{}` creates a `_CompactLinkedHashSet<int>`).

The fallback strategy of searching for the type in the type tree can be made even more efficient by registering common `dart:core` types first (like `Iterable`, which in turn contains `List` and `Set` first).

> I'm still not sure whether I like the idea to create adapters for primitives and Lists. The reason is that the `BinaryReader` and `BinaryWriter` methods can access the byte buffer directly which improves performance.

Personally, I really like the cleanliness of the architecture where every type has a corresponding adapter instead of some types having hardcoded behavior.
Though I do get that performance is a top priority for you.

But I'm not sure I completely get your point here – adapters have no overhead whatsoever.
In fact, with the current constant type lookup before the tree traversal, looking up the `AdapterForList` and `AdapterForSet` is more efficient than before.

I'm not sure about the performance as I'm not well-versed with Dart performance – as you said, the docs on that are virtually non-existent. :trollface: 
But if the current `BinaryWriter` and `BinaryReader` interfaces don't allow the most efficient operating on data, maybe the API could be changed so it's possible to delegate operations directly to a `ByteBuffer`? So, instead of using `Uint8List` as the exchange format, they might have some operations like `writeUint8(value)`, which in turn just calls the corresponding call at the correct offset on an internal `ByteBuffer`.

> I don't see why anyone would want to override the default behavior for storing primitives.

That's not what I meant, sorry if I didn't make myself clear.
I was talking about how it would make it easier for _us_ as framework implementers to register these special types.

> A small note on using `ByteData` and `Uint8List`. Creating a new instance of either of them is rather expensive (even `Uint8List.view()`).

I do like the simplicity of an API that just implements `writeBytes(List<int> bytes)`, but I'll experiment with broadening the API surface to a few functions to yield better performance by directly acting on the `ByteBuffer`.

**Comment by leisim** [comment]
> But I'm not sure I completely get your point here – adapters have no overhead whatsoever.

I suspect Dart can optimize the `if (value is SomeType) else if ...` code. But I have not tested it yet and need to run some microbenchmarks. I fully agree that the adapter approach is cleaner.

> I do like the simplicity of an API that just implements `writeBytes(List<int> bytes)`

Yes I agree but in that case I don't think it is worth it.

> I'll experiment with broadening the API surface to a few functions to yield better performance by directly acting on the `ByteBuffer`

I've used "package private" extensions to hide internal methods for the `HiveObject` and it worked quite well.

I will also create a prototype of this idea, then we can compare our solutions and maybe settle with the optimal compromise. Currently I'm working on a new project so I will probably need until late february to implement it.

**Comment by marcelgarus** [comment]
🎉 I'm pretty much done with my [serializer prototype](https://github.com/dart-portals/binary/).

You can now register several adapters for a single type. That allows for some nifty optimizations like encoding some `int`s as `uint8` and others as `int16` based on their value.
Admittedly, I went a little overboard with the optimization of the binary format. For example, the list `[1, 6, 10000]` gets encoded using the `AdapterForPrimitiveList.short(AdapterForUint16())`, which is really optimized for this kind of list – because the list is short and doesn't contain `null`s, only one byte saves the size and the rest of the bytes save raw uint16s. Some of the currently over 100 adapters should probably be removed in favor of a less efficient but faster serialization – I guess that's the typical trade-off between time and space efficiency.

Also, I have to say I really like the two bytes for type ids. They give us a lot of room for external packages registering adapters. That means developers could mix and match packages if they all use non-overlapping type ids. A [table of reserved type ids](https://github.com/dart-portals/binary/blob/master/table_of_type_ids.md) could ensure that's the case.

**Comment by marcelgarus** [comment]
I really like the current implementation of <kbd>binary</kbd> as it provides easy-to-use type-safe serializing/deserializing. I'll probably make it a little more time-efficient.
I suppose in the new architecture, the communication between the backend and frontend probably happens over some standardized protocol like JSON or ProtoBuf. But at some point, a serialization of Dart objects still needs to happen. So, how should we proceed with this issue?

**Comment by leisim** [comment]
> I suppose in the new architecture, the communication between the backend and frontend probably happens over some standardized protocol like JSON or ProtoBuf.

I think the `BinaryReader` and `BinaryWriter` are one of the few pieces that will actually remain in the new version.
Since Dart does not allow sending or receiving objects to C or Rust, they will need to be serialized. I have come up with a format that allows random accessing specific fields without reading the whole object. This should make queries super fast.

I did not have time to benchmark the `binary` package yet so I don't know if it is a suitable replacement for `BinaryReader` / `BinaryWriter`. I would leave this issue open until then.

**Comment by marcelgarus** [comment]
I ended up inventing a whole new serialization format that is self-describing like JSON. That means it supports inspecting objects created with adapters unknown to us as well as parsing objects in other languages like Rust. These primitive blocks are also designed to be extensible over time.

Here's the repository: https://github.com/marcelgarus/tape
And here's the kind-of specification/thoughts about the basic building blocks: https://github.com/marcelgarus/tape/issues/27

The other layers (adapters, registry, etc) still need to be fleshed out.
I also plan to create more tooling and infrastructure that are needed for a great serializer, like drop-in-support for types of other packages and tools helping with annotating objects. Finally, because it's based on annotating existing classes (like Hive is currently), it also supports <kbd>freezed</kbd> (at least theoretically, I still need to implement that part).
